src/charm.py

```diff
@@ -0,0 +1,210 @@
+            logger.info(f"Received servers {servers}")
+            if not servers:
+                self.unit.status = BlockedStatus("Waiting for etcd")
+                return
+            etcd_cfg.servers = servers
+        backend_cfg = BackendConfig(
+            port_range=self.config["port_range"],
+            lxd_project=self.config["lxd_project"],
+            force_tls12=self.config["force_tls12"],
+            use_network_acl=self.config["use_network_acl"],
+        )
+        if self.config["metrics_server"]:
+            backend_cfg.metrics_server = f"influxdb:{self.config['metrics_server']}"
+
+        metrics_cfg = PrometheusConfig(
+            ip=self.private_ip,
+            port=int(self.config["prometheus_target_port"]),
+            tls_cert_path=self.config["prometheus_tls_cert_path"],
+            tls_key_path=self.config["prometheus_tls_key_path"],
+            basic_auth_username=self.config["prometheus_basic_auth_username"],
+            basic_auth_password=self.config["prometheus_basic_auth_password"],
+            extra_labels=self.config["prometheus_extra_labels"],
+            metrics_path=self.config["prometheus_metrics_path"],
+        )
+        cfg = ServiceConfig(
+            ip=self.private_ip,
+            port=int(self.config["port"]),
+            log_level=self.config["log_level"],
+            metrics=metrics_cfg,
+            backend=backend_cfg,
+            store=etcd_cfg,
+        )
+        self._snap.configure(cfg)
+        if self.config["location"]:
+            self._snap.set_location(self.config["location"], self.config["port"])
+        self.unit.set_ports(int(self.config["port"]))
+        self.unit.status = ActiveStatus()
+
+    def _on_lxd_integrator_joined(self, event: RelationJoinedEvent):
+        cert, key = self._generate_selfsigned_cert(self.public_ip, self.public_ip, self.private_ip)
+        self._snap.setup_lxd(cert=cert, key=key)
+        relation_data = event.relation.data[self.unit]
+        relation_data["client_certificates"] = json.dumps([cert.decode("utf-8")])
+
+    def _generate_selfsigned_cert(self, hostname, public_ip, private_ip) -> tuple[bytes, bytes]:
+        if not hostname:
+            raise Exception("A hostname is required")
+
+        if not public_ip:
+            raise Exception("A public IP is required")
+
+        if not private_ip:
+            raise Exception("A private IP is required")
+
+        ca_key = generate_private_key(key_size=4096)
+        ca_cert = generate_ca(ca_key, hostname)
+
+        key = generate_private_key(key_size=4096)
+        csr = generate_csr(
+            private_key=key,
+            subject=hostname,
+            sans_dns=[public_ip, private_ip, hostname],
+            sans_ip=[public_ip, private_ip],
+        )
+        cert = generate_certificate(csr=csr, ca=ca_cert, ca_key=ca_key)
+        return cert, key
+
+    def _on_rest_api_joined(self, event: RelationJoinedEvent):
```

Comment:
Something we can refactor later, but should all API endpoint things go into a separate class?

Reply:
Yes we can move these to separate endpoint classes later on.